File: refactor.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const filepath = path.join(__dirname, "components", "UseCasesGrid.tsx");
const outdir = path.join(__dirname, "components", "use-cases");

if (!fs.existsSync(outdir)) {
    fs.mkdirSync(outdir, { recursive: true });
}

let content = fs.readFileSync(filepath, 'utf8');

const cases = [
    ["MorningRush", "\\{\\/\\* Use Case 1: Morning Rush \\*\\/\\}", "\\{\\/\\* Use Case 2: Rainy Day Protocol \\*\\/\\}"],
    ["RainyDay", "\\{\\/\\* Use Case 2: Rainy Day Protocol \\*\\/\\}", "\\{\\/\\* Use Case 3: The 'Biryani Index' \\*\\/\\}"],
    ["BiryaniIndex", "\\{\\/\\* Use Case 3: The 'Biryani Index' \\*\\/\\}", "\\{\\/\\* Use Case 4: Vendor Auto-Negotiation \\*\\/\\}"],
    ["VendorNegotiation", "\\{\\/\\* Use Case 4: Vendor Auto-Negotiation \\*\\/\\}", "\\{\\/\\* Use Case 5: Staff Language Bridge \\*\\/\\}"],
    ["LanguageBridge", "\\{\\/\\* Use Case 5: Staff Language Bridge \\*\\/\\}", "\\{\\/\\* Use Case 6: The 'Bandh' Shield \\*\\/\\}"],
    ["BandhShield", "\\{\\/\\* Use Case 6: The 'Bandh' Shield \\*\\/\\}", "\\{\\/\\* Use Case 7: Smart-Menu Engineering \\*\\/\\}"],
    ["SmartMenu", "\\{\\/\\* Use Case 7: Smart-Menu Engineering \\*\\/\\}", "\\{\\/\\* Use Case 8: "],
    ["ExitStrategy", "\\{\\/\\* Use Case 18: The 'Exit Strategy' \\*\\/\\}", "\\{\\/\\* Divider \\*\\/\\}"]
];

const imports = `import React from 'react';
import { motion } from 'framer-motion';
import { Link } from 'react-router-dom';
import { PageRoute } from '../../types';
import { 
  Coffee, CloudRain, TrendingUp, Truck, Mic, ShieldAlert, 
  Menu, UserCheck, Battery, MessageSquare, AlertTriangle, Search, FileText, 
  Users, Award, Recycle, MapPin, DollarSign, Brain, CloudLightning, Smartphone, Clock, ArrowDownRight, ArrowRight
} from 'lucide-react';

interface Slide {
  id: number;
  icon: any;
  color: string;
  bgColor: string;
  borderColor: string;
  title: string;
  description: string;
  tag: string;
  stat: string;
}

interface UseCaseProps {
  slide: Slide;
}

export const ExpandedUseCases: React.FC<{ expandedSlides: Slide[] }> = ({ expandedSlides }) => {
  return (
    <div className="space-y-32 mb-32">
        <UseCaseMorningRush slide={expandedSlides[0]} />
        <UseCaseRainyDay slide={expandedSlides[1]} />
        <UseCaseBiryaniIndex slide={expandedSlides[2]} />
        <UseCaseVendorNegotiation slide={expandedSlides[3]} />
        <UseCaseLanguageBridge slide={expandedSlides[4]} />
        <UseCaseBandhShield slide={expandedSlides[5]} />
    </div>
  );
};
`;

let topImports = [];

// For safety, I will not split into 18 components automatically right now, that is extremely complex string matching.
// Instead I will just extract the entire block from "Use Case 1: Morning Rush" to "Divider" into one huge new file "ExpandedUseCases.tsx".
// Then replace that block in UseCasesGrid.tsx with `<ExpandedUseCases expandedSlides={expandedSlides} />`.

const startMatch = content.match(/\{\/\* Use Case 1: Morning Rush \*\/\}/);
const endMatch = content.match(/\{\/\* Divider \*\/\}/);

if (startMatch && endMatch) {
    const startIdx = startMatch.index;
    const endIdx = endMatch.index;
    const block = content.substring(startIdx, endIdx).trim();
    
    // We rewrite ExpandedUseCases.tsx
    const componentCode = `${imports}

export const ExpandedUseCases: React.FC<{ expandedSlides: Slide[] }> = ({ expandedSlides }) => {
  return (
    <div className="space-y-32 mb-32">
      ${block}
    </div>
  );
};
`;
    // We remove the old ExpandedUseCases if it exists in `imports` (just rewrite).
    
    // Fix the `imports` variable slightly above since I duplicated it.
    const cleanComponentCode = componentCode.replace(/export const ExpandedUseCases[\s\S]*?;\n};\n/, ''); // remove duplicate

    fs.writeFileSync(path.join(outdir, "ExpandedUseCases.tsx"), cleanComponentCode, "utf8");
    
    let newContent = content.substring(0, startIdx) + "<ExpandedUseCases expandedSlides={expandedSlides} />\n\n        " + content.substring(endIdx);
    
    // Inject import into newContent
    const insertPos = newContent.indexOf("const slides = [");
    newContent = newContent.substring(0, insertPos) + "import { ExpandedUseCases } from './use-cases/ExpandedUseCases';\n" + newContent.substring(insertPos);
    
    fs.writeFileSync(filepath, newContent, "utf8");
    console.log("Success! Refactored " + filepath);
} else {
    console.log("Could not find the bounds for the entire expanded slides block.");
}

File: refactor.py
import re
import os

filepath = "components/UseCasesGrid.tsx"
outdir = "components/use-cases"
os.makedirs(outdir, exist_ok=True)

with open(filepath, "r", encoding="utf-8") as f:
    content = f.read()

# Define the boundaries of the use cases we want to extract
cases = [
    ("MorningRush", r"\{\/\* Use Case 1: Morning Rush \*\/\}", r"\{\/\* Use Case 2: Rainy Day Protocol \*\/\}"),
    ("RainyDay", r"\{\/\* Use Case 2: Rainy Day Protocol \*\/\}", r"\{\/\* Use Case 3: The 'Biryani Index' \*\/\}"),
    ("BiryaniIndex", r"\{\/\* Use Case 3: The 'Biryani Index' \*\/\}", r"\{\/\* Use Case 4: Vendor Auto-Negotiation \*\/\}"),
    ("VendorNegotiation", r"\{\/\* Use Case 4: Vendor Auto-Negotiation \*\/\}", r"\{\/\* Use Case 5: Staff Language Bridge \*\/\}"),
    ("LanguageBridge", r"\{\/\* Use Case 5: Staff Language Bridge \*\/\}", r"\{\/\* Use Case 6: The 'Bandh' Shield \*\/\}"),
    ("BandhShield", r"\{\/\* Use Case 6: The 'Bandh' Shield \*\/\}", r"\{\/\* Compact Use Cases Grid \*\/\}")
]

imports = """import React from 'react';
import { motion } from 'framer-motion';
import { Link } from 'react-router-dom';
import { PageRoute } from '../../types';
import { 
  Coffee, CloudRain, TrendingUp, Truck, Mic, ShieldAlert, 
  Menu, UserCheck, Battery, MessageSquare, AlertTriangle, Search, FileText, 
  Users, Award, Recycle, MapPin, DollarSign, Brain, CloudLightning, Smartphone, Clock, ArrowDownRight, ArrowRight
} from 'lucide-react';

interface Slide {
  id: number;
  icon: any;
  color: string;
  bgColor: string;
  borderColor: string;
  title: string;
  description: string;
  tag: string;
  stat: string;
}

interface UseCaseProps {
  slide: Slide;
}
"""

components_dict = {}

for name, start_re, end_re in cases:
    # Need to find the blocks.
    match_start = re.search(start_re, content)
    match_end = re.search(end_re, content)
    if match_start and match_end:
        start_idx = match_start.start()
        end_idx = match_end.start()
        
        block = content[start_idx:end_idx].strip()
        
        # We need to replace `expandedSlides[index]` with `slide`
        # Because we'll pass slide as a prop.
        block = re.sub(r'expandedSlides\[\d+\]', 'slide', block)
        
        component_code = f"""{imports}
export const UseCase{name}: React.FC<UseCaseProps> = ({{ slide }}) => {{
  return (
    {block}
  );
}};
"""
        with open(os.path.join(outdir, f"UseCase{name}.tsx"), "w", encoding="utf-8") as f:
            f.write(component_code)
            
        components_dict[name] = block
        
print("Extracted Use Cases.")

# Now replace those chunks in the main file
new_content = content
top_imports = []
replace_str = ""

for name, _, _ in cases:
    top_imports.append(f"import {{ UseCase{name} }} from './use-cases/UseCase{name}';")
    # Instead of deleting, just replacing the giant block in new_content
    # The `cases` specify start and end bounds exactly. This is tricky due to re substitutions.

# To be safe, just replace the entire Expanded Top 3 Use Cases / 6 bounds.
match_first_case = re.search(cases[0][1], new_content)
match_last_bound = re.search(cases[-1][2], new_content)

if match_first_case and match_last_bound:
    start_cut = match_first_case.start()
    end_cut = match_last_bound.start()
    
    components_render = """
          <UseCaseMorningRush slide={expandedSlides[0]} />
          <UseCaseRainyDay slide={expandedSlides[1]} />
          <UseCaseBiryaniIndex slide={expandedSlides[2]} />
          <UseCaseVendorNegotiation slide={expandedSlides[3]} />
          <UseCaseLanguageBridge slide={expandedSlides[4]} />
          <UseCaseBandhShield slide={expandedSlides[5]} />
          
          """
          
    latest_content = new_content[:start_cut] + components_render + new_content[end_cut:]
    
    # insert imports at top (after typical imports)
    insert_pos = latest_content.find("const slides = [")
    latest_content = latest_content[:insert_pos] + "\n".join(top_imports) + "\n\n" + latest_content[insert_pos:]
    
    with open(filepath, "w", encoding="utf-8") as f:
        f.write(latest_content)
        
    print("Replaced main file content successfully. Size went from len", len(content), "to", len(latest_content))
else:
    print("Could not find boundaries for cut.")
